combainer/serf.go

```diff
@@ -0,0 +1,327 @@
+		leaderCh:   make(chan bool, 1),
+
+		raftAdvertiseIP: raftAdvertiseIP,
+		store:           NewFSMStore(),
+		log:             log,
+		cache:           cache,
+		repo:            repo,
+		config:          &cfg,
+	}
+	return c, nil
+}
+
+// Cluster is wrapper for access cluster members
+type Cluster struct {
+	Name string
+
+	// eventCh is used to receive events from the serf cluster
+	eventCh chan serf.Event
+	serf    *serf.Serf
+
+	// reconcileCh is used to pass events from the serf handler
+	// into the leader manager. Mostly used to handle when servers
+	// join/leave from the region.
+	reconcileCh chan serf.Member
+
+	shutdownCh chan struct{}
+	leaderCh   chan bool
+
+	raftAdvertiseIP net.IP
+	raft            *raft.Raft
+	transport       *raft.NetworkTransport
+	raftStore       *raftboltdb.BoltStore
+	raftConfig      *raft.Config
+
+	store          *FSMStore
+	updateInterval time.Duration
+	log            *logrus.Entry
+	cache          cache.Cache
+	repo           configs.Repository
+	config         *configs.ClusterConfig
+}
+
+// Bootstrap is used to attempt join to existing serf cluster.
+// and bootstrap the Raft agent using cluster as FSM.
+// Updates leadership are returned on leaderCh,
+// leader dispatch new configs every interval time.
+func (c *Cluster) Bootstrap(initHosts []string, interval time.Duration) error {
+	c.log.Infof("Bootstrap cluster, connect Serf nodes: %s", initHosts)
+CONNECT:
+	n, err := c.serf.Join(initHosts, true)
+	if n > 0 {
+		c.log.Infof("Combainer joined to cluster: %d nodes", n)
+	}
+	// NOTE: doc from serf.Join
+	// Join joins an existing Serf cluster. Returns the number of nodes
+	// successfully contacted. The returned error will be non-nil only in the
+	// case that no nodes could be contacted.
+	if err != nil {
+		c.log.Errorf("Combainer error joining to cluster: %d nodes", n)
+		time.Sleep(interval)
+		goto CONNECT
+	}
+
+	c.log.Info("Create raft transport")
+	c.transport, err = raft.NewTCPTransport(
+		net.JoinHostPort(c.config.BindAddr, strconv.Itoa(c.config.RaftPort)),
+		&net.TCPAddr{IP: c.raftAdvertiseIP, Port: c.config.RaftPort},
+		raftPool,
+		raftTimeout,
+		c.log.Logger.Writer(),
+	)
+	if err != nil {
+		return errors.Wrap(err, "tcp transport failed")
+	}
+
+	var peersAddrs []string
+	for _, m := range c.Members() {
+		addr := net.JoinHostPort(m.Addr.String(), strconv.Itoa(c.config.RaftPort))
+		peersAddrs = append(peersAddrs, addr)
+	}
+	c.log.Infof("Set raft peers %v", peersAddrs)
+	raftPeers := raft.NewJSONPeers(c.config.RaftStateDir, c.transport)
+	if err = raftPeers.SetPeers(peersAddrs); err != nil {
+		return err
+	}
+
+	c.log.Info("Create snapshot store")
+	snapshots, err := raft.NewFileSnapshotStore(
+		c.config.RaftStateDir, retainRaftSnapshot, c.log.Logger.Writer(),
+	)
+	if err != nil {
+		return err
+	}
+
+	c.log.Info("Create raft log store")
+	boltStore, err := raftboltdb.NewBoltStore(filepath.Join(c.config.RaftStateDir, "raft.db"))
+	if err != nil {
```

Comment:
defer shutdown https://github.com/combaine/combaine/pull/35/commits/c7d02fa978b4cf9e12a13c063ddaa29bc4ca6608#diff-8bd6e1d0e6d154fd6c35151cb28d33b2R101
and release all resources there.